src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java

```diff
@@ -15,15 +15,179 @@
+    }
+
+    try {
+      Awaitility
+        .await().with()
+        .pollInterval(pollInterval)
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will ba handled using httpStatusCode below.
+    }
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
+    if (config.getFormat().equals(PageFormat.BLOB)) {
+      responseBytes = Base64.encodeBase64(responseBytes);
+    }
+    String responseBody = new String(responseBytes, StandardCharsets.UTF_8);
+
+    HttpErrorHandlingStrategy postRetryStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode)
+      .getAfterRetryStrategy();
+
+    String errorMessage = String.format("Fetching from url '%s' returned status code '%d' and body '%s'",
+                                        nextPageUrl, httpStatusCode, responseBody);
+
+    switch (postRetryStrategy) {
+      case SUCCESS:
+        break;
+      case FAIL:
+        throw new IllegalStateException(errorMessage);
+      case SKIP:
+      case SEND_TO_ERROR:
+        if (!this.supportsSkippingPages()) {
+          throw new IllegalStateException(String.format(
+            "Pagination type '%s', does not support 'skip' and 'send to error' error handling.",
+            config.getPaginationType()));
+        }
+        LOG.warn(errorMessage);
+        // this will be handled by PageFactory. Here no handling is needed.
+        break;
+      default:
+        throw new IllegalArgumentException(String.format("Unexpected http error handling: '%s'", postRetryStrategy));
+    }
+
+    BasePage page = PageFactory.createInstance(config, responseBody, postRetryStrategy);
+    nextPageUrl = getNextPageUrl(responseBody, response, page);
+
+    LOG.debug("Next Page Url is '{}'", nextPageUrl);
+    //LOG.info("Schema is: {}", config.getSchemaString());
+
+    return page;
+  }
+
+  /**
+   * @return true if page still has elements to iterate. Otherwise it will load next page.
+   * False if no more pages to load or the page loaded has no elements.
+   */
+  protected boolean ensurePageIterable() {
+    try {
+      if (page == null || !page.hasNext()) {
+        page = getNextPage();
+      }
+
+      return (page != null && page.hasNext());
+    } catch (IOException | InterruptedException e) {
+      throw new IllegalStateException("Failed to the load page", e);
+    }
+  }
+
+  @Override
+  public PageEntry next() {
+    if (!ensurePageIterable()) {
+      throw new NoSuchElementException("No more pages to load.");
+    }
+
+    return new PageEntry(httpStatusCode, page.next());
+  }
+
+  @Override
+  public boolean hasNext() {
+    return ensurePageIterable();
+  }
+
+  @Override
+  public void close() throws IOException {
```

Comment:
does not guarantee that both resources will close in case of exception

Reply:
done